src/storage-controller/src/lib.rs

```diff
@@ -1134,6 +1137,88 @@ where
         Ok(())
     }
 
+    /// Create the sinks described by the `ExportDescription`.
+    async fn update_export_connection(
+        &mut self,
+        exports: BTreeMap<GlobalId, StorageSinkConnection>,
+    ) -> Result<(), StorageError> {
+        for (id, connection) in exports {
+            let export = self.export_mut(id).expect("export exists");
+            let current_sink = export.description.sink.clone();
+            export.description.sink.connection = connection;
+
+            // Ensure compatibility
+            current_sink.alter_compatible(id, &export.description.sink)?;
+
+            // Get export data
+            let value = MetadataExportFetcher::get_stash_collection()
+                .peek_key_one(&mut self.stash, id.into_proto())
+                .await?
+                .expect("known to exist");
+
+            let mut durable_export_data = DurableExportMetadata::from_proto(value)
+                .map_err(|e| StorageError::IOError(e.into()))?;
+
+            // Get updated export state
+            let export = self.export(id).expect("export exists").clone();
+
+            let from_collection = self.collection(export.description.sink.from)?;
+            let from_storage_metadata = from_collection.collection_metadata.clone();
+
+            // We've already installed our read holds on this collection, so we
+            // need to get its read capabilities (rather than its implied
+            // capability, which we would if it were a new export).
+            let from_read_capability = from_collection.read_capabilities.frontier().to_owned();
```

Comment:
The frontier you get by calling `frontier()` on the `read_capabilities` accumulation is the minimum across *all* holds and not just the hold the sink put on the from collection. So it's entirely possible that you get a small `as_of` here that you send down to the cluster and then whoever was holding that `as_of` to that low value to release their hold, rendering your command invalid.

Reply:
This was an oversight on my end––I'm aware of how mutable antichains work, but had overlooked that we had the data we needed already tracked in the export state. Fixed on my next push.

Reply:
Fixed